src/Microsoft.OData.Core/Json/JsonValueUtilsAsync.cs

```diff
@@ -0,0 +1,468 @@
+                    // Otherwise copy to the buffer and go on from there.
+                    if (substrLength >= bufferLength)
+                    {
+                        inputString.CopyTo(currentIndex, buffer.Value, 0, bufferLength);
+                        await writer.WriteAsync(buffer.Value, 0, bufferLength).ConfigureAwait(false);
+                        currentIndex += bufferLength;
+                    }
+                    else
+                    {
+                        WriteSubstringToBuffer(inputString, ref currentIndex, buffer.Value, ref bufferIndex, substrLength);
+                    }
+                }
+
+                // Write escaped string to buffer
+                WriteEscapedStringToBuffer(writer, inputString, ref currentIndex, buffer.Value, ref bufferIndex, stringEscapeOption);
+
+                // write any remaining chars to the writer
+                if (bufferIndex > 0)
+                {
+                    await writer.WriteAsync(buffer.Value, 0, bufferIndex).ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Escapes and writes a character array to a writer.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="inputArray">Character array to write.</param>
+        /// <param name="inputArrayOffset">How many characters to skip in the input array.</param>
+        /// <param name="inputArrayCount">How many characters to write from the input array.</param>
+        /// <param name="stringEscapeOption">The string escape option.</param>
+        /// <param name="bufferPool">Character buffer pool.</param>
+        internal static async Task WriteEscapedCharArrayAsync(this TextWriter writer, char[] inputArray, int inputArrayOffset, int inputArrayCount, ODataStringEscapeOption stringEscapeOption, ICharArrayPool bufferPool)
+        {
+            int bufferIndex = 0;
+            char[] buffer = null;
+
+            buffer = BufferUtils.InitializeBufferIfRequired(bufferPool, buffer);
+            
+            WriteEscapedCharArrayToBuffer(writer, inputArray, ref inputArrayOffset, inputArrayCount, buffer, ref bufferIndex, stringEscapeOption);
+
+            // write remaining bytes in buffer
+            if (bufferIndex > 0)
+            {
+                await writer.WriteAsync(buffer, 0, bufferIndex).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Write the string value with quotes.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="text">String value to be written.</param>
+        private static async Task WriteQuotedAsync(this TextWriter writer, string text)
+        {
+            await writer.WriteAsync(
+                JsonConstants.QuoteCharacter + text + JsonConstants.QuoteCharacter)

```

Comment:
>JsonConstants.QuoteCharacter + text + JsonConstants.QuoteCharacter [](start = 16, length = 66)

Consider using String.Concat(JsonConstants.QuoteCharacter, text, JsonConstants.QuoteCharacter) instead of JsonConstants.QuoteCharacter + text + JsonConstants.QuoteCharacter. I think the compiler might automatically optimize this case by generating the concat, rather than do two allocations, but I'm not sure.